File: parser_.py
from errors import Error
from tokens import Token, TokenType
from nodes import Node

class Parser:
    def __init__(self, file, tokens):
        self.file = file
        self.tokens = iter(tokens)
        self.advance()

    def raise_error(self, token):
        raise Error(self.file, token.line, 'syntax error')

    def advance(self):
        self.current_token = next(self.tokens)

    def parse(self):
        if self.current_token.type == TokenType.EOF:
            return Node(self.current_token.line, ('empty',))

        result = self.statements()

        if self.current_token.type != TokenType.EOF:
            self.raise_error(self.current_token)

        return result

    def statements(self):
        line = self.current_token.line
        statements = []

        statement = self.expr()
        statements.append(statement)

        while self.current_token != None and self.current_token.type == TokenType.NEWLINE:
            self.advance()
            statements.append(self.expr())
            
        return Node(line, ('statements', *statements))

    def expr(self):
        return self.additive_expr()
    
    def additive_expr(self):
        result = self.multiplicative_expr()

        while self.current_token != None and self.current_token.type in (TokenType.PLUS, TokenType.MINUS):
            if self.current_token.type == TokenType.PLUS:
                self.advance()
                result = Node(result.line, ('add', result, self.multiplicative_expr()))
            elif self.current_token.type == TokenType.MINUS:
                self.advance()
                result = Node(result.line, ('subtract', result, self.multiplicative_expr()))

        return result

    def multiplicative_expr(self):
        result = self.factor()

        while self.current_token != None and self.current_token.type in (TokenType.MULTIPLY, TokenType.DIVIDE, TokenType.MOD):
            if self.current_token.type == TokenType.MULTIPLY:
                self.advance()
                result = Node(result.line, ('multiply', result, self.factor()))
            elif self.current_token.type == TokenType.DIVIDE:
                self.advance()
                result = Node(result.line, ('divide', result, self.factor()))
            elif self.current_token.type == TokenType.MOD:
                self.advance()
                result = Node(result.line, ('mod', result, self.factor()))

        return result

    def factor(self):
        token = self.current_token

        if token.type == TokenType.LPAREN:
            self.advance()
            result = self.expr()

            if self.current_token.type != TokenType.RPAREN:
                self.raise_error(self.current_token)
            
            self.advance()
            return result

        elif token.type == TokenType.NUMBER:
            self.advance()
            return Node(token.line, ('number', token.value))

        elif token.type == TokenType.PLUS:
            self.advance()
            return Node(token.line, ('plus', self.factor()))
        
        elif token.type == TokenType.MINUS:
            self.advance()
            return Node(token.line, ('minus', self.factor()))
        
        self.raise_error(self.current_token)

File: Parser.js
var moo = require("moo");
var LangError = require("./LangError.js");

class Parser {
    constructor(tokens) {
        this.tokens = tokens.filter((token) => token.type !== "WS");
        this.idx = -1;
        this.advance();
    }

    raiseError(token) {
        var msg = moo.compile({}).formatError(token, "invalid syntax").split("\n")[0].slice(0, -1);
        new LangError(msg, token.offset).raise();
    }

    advance() {
        this.idx += 1;
        if (this.idx >= this.tokens.length) {
            this.currentToken = this.tokens[this.tokens.length - 1];
        } else {
            this.currentToken = this.tokens[this.idx]
        }
    }

    parse() {
        var result = this.expr();
        this.advance();
        if (this.currentToken.type !== "eof") {
            this.raiseError(this.currentToken);
        }

        return result;
    }

    expr() {
        return this.addExpr();
    }

    addExpr() {
        var left = this.mulExpr();
        while (this.currentToken.type === "plus" || this.currentToken.type === "minus") {
            var token = this.currentToken;

            this.advance();
            var right = this.mulExpr();
            
            if (token.type === "plus") {
                left = {
                    type: "plus",
                    value: [left, right],
                    offset: left.offset
                };
            } else {
                left = {
                    type: "minus",
                    value: [left, right],
                    offset: left.offset
                };
            }
        }

        return left;
    }
    
    mulExpr() {
        var left = this.atom();
        while (this.currentToken.type === "mul" || this.currentToken.type === "div" || this.currentToken.type === "mod") {
            var token = this.currentToken;

            this.advance();
            var right = this.atom();
            
            if (token.type === "mul") {
                left = {
                    type: "mul",
                    value: [left, right],
                    offset: left.offset
                };
            } else if(token.type === "div") {
                left = {
                    type: "div",
                    value: [left, right],
                    offset: left.offset
                };
            } else {
                left = {
                    type: "mod",
                    value: [left, right],
                    offset: left.offset
                };
            }
        }

        return left;
    }
    
    atom() {
        var token = this.currentToken;

        if (token.type === "int") {
            this.advance();
            return {
                type: "int",
                value: token.value,
                offset: token.offset
            };
        } else if (token.type === "float") {
            this.advance();
            return {
                type: "float",
                value: token.value,
                offset: token.offset
            };
        } else if (token.type === "plus") {
            this.advance();
            var atom = this.atom();

            return {
                type: "uplus",
                value: [atom],
                offset: token.offset
            };
        } else if (token.type === "minus") {
            this.advance();
            var atom = this.atom();

            return {
                type: "uminus",
                value: atom,
                offset: token.offset
            };
        } else if (token.type === "lparen") {
            this.advance();
            var expr = this.expr();
            if (this.currentToken.type !== "rparen") {
                this.raiseError(this.currentToken);
            }

            this.advance();
            return expr;
        } else {
            this.raiseError(this.currentToken);
        }
    }
}

module.exports = Parser;